community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -69,14 +67,64 @@ case class ExpandIntoPipe(source: Pipe,
     }
   }
 
+  private def iterateRelationships(row: ExecutionContext, state: QueryState, fromNode: Node, toNode: Node) = {
+
+    //check degree and iterate from the node with smaller degree
+    val fromDegree = state.query.nodeGetDegree(fromNode.getId, dir)
+    val toDegree = state.query.nodeGetDegree(toNode.getId, dir.reverse())
+    val (start, end, relationships) = if (fromDegree < toDegree)
+      (fromNode, toNode, state.query.getRelationshipsForIds(fromNode, dir, types.types(state.query)))
```

Comment:
what happens here if we don't have a type? or if there are multiple? is that resolved elsewhere already?